# Data Profiling: Canadian Job Market Analysis


## Column Evaluation for Silver Layer

### 1) job_title

In [0]:
%sql
SELECT DISTINCT job_title, COUNT(*) FROM workspace.dev_jobmkt.stg_job_postings GROUP BY 1 ORDER BY 2 DESC

job_title,COUNT(*)
cook,1291
food service supervisor,865
continuing care assistant,659
line cook,583
administrative assistant,531
cashier,481
restaurant manager,471
carpenter,439
sales associate,404
delivery driver,396


### 2) noc16_code

In [0]:
%sql
SELECT DISTINCT noc16_code, COUNT(*) FROM workspace.dev_jobmkt.stg_job_postings GROUP BY 1 ORDER BY 2 DESC

noc16_code,COUNT(*)
6322,2483
6711,1945
6421,1797
7511,1371
6311,1312
0631,1127
8431,1080
4411,1029
4412,980
6731,973


### 3) noc16_code_name

In [0]:
%sql
SELECT DISTINCT noc16_code_name, COUNT(*) FROM workspace.dev_jobmkt.stg_job_postings GROUP BY 1 ORDER BY 2 DESC

noc16_code_name,COUNT(*)
Cooks,2483
"Food counter attendants, kitchen helpers and related support occupations",1945
Retail salespersons,1797
Transport truck drivers,1371
Food service supervisors,1312
Restaurant and food service managers,1127
General farm workers,1080
Home child care providers,1029
"Home support workers, housekeepers and related occupations",980
Light duty cleaners,973


### 4) noc21_code

In [0]:
%sql
SELECT DISTINCT noc21_code, COUNT(*) FROM workspace.dev_jobmkt.stg_job_postings GROUP BY 1 ORDER BY 2 DESC

noc21_code,COUNT(*)
63200,2483
65201,1945
64100,1810
73300,1363
62020,1312
60030,1127
65310,1052
44100,1029
33102,927
14101,922


### 5) noc21_code_name

In [0]:
%sql
SELECT DISTINCT noc21_code_name, COUNT(*) FROM workspace.dev_jobmkt.stg_job_postings GROUP BY 1 ORDER BY 2 DESC

noc21_code_name,COUNT(*)
Cooks,2483
"Food counter attendants, kitchen helpers and related support occupations",1945
Retail salespersons and visual merchandisers,1810
Transport truck drivers,1363
Food service supervisors,1312
Restaurant and food service managers,1127
Light duty cleaners,1052
Home child care providers,1029
"Nurse aides, orderlies and patient service associates",927
Receptionists,922


### 6) posting_date

In [0]:
%sql
SELECT DISTINCT posting_date, COUNT(*) FROM workspace.dev_jobmkt.stg_job_postings GROUP BY 1 ORDER BY 2 DESC

posting_date,COUNT(*)
2026-08-04,2714
2026-08-19,2655
2026-08-10,2594
2026-08-06,2592
2026-08-05,2586
2026-08-13,2560
2026-08-18,2553
2026-08-20,2505
2026-08-17,2462
2026-08-11,2459


### 7) vacancy_count

In [0]:
%sql
SELECT DISTINCT vacancy_count, COUNT(*) FROM workspace.dev_jobmkt.stg_job_postings GROUP BY 1 ORDER BY 2 DESC

vacancy_count,COUNT(*)
1,43192
2,5383
3,1676
4,922
5,830
10,419
6,367
8,195
20,111
7,104


### 8) education_los

In [0]:
%sql
SELECT DISTINCT education_los, COUNT(*) FROM workspace.dev_jobmkt.stg_job_postings GROUP BY 1 ORDER BY 2 DESC

education_los,COUNT(*)
null,30999
Secondary (high) school graduation certificate,10595
"No degree, certificate or diploma",5964
College/CEGEP,1400
Other trades certificate or diploma,1253
Bachelor's degree,1212
"College, CEGEP or other non-university certificate or diploma from a program of 1 year to 2 years",1208
"College, CEGEP or other non-university certificate or diploma from a program of 3 months to less than 1 year",478
Registered Apprenticeship certificate,359
Master's degree,137


### 9) experience_level

In [0]:
%sql
SELECT DISTINCT experience_level, COUNT(*) FROM workspace.dev_jobmkt.stg_job_postings GROUP BY 1 ORDER BY 2 DESC

experience_level,COUNT(*)
null,30999
1 year to less than 2 years,6348
No experience,5821
2 years to less than 3 years,3449
3 years to less than 5 years,2165
1 to less than 7 months,1974
7 months to less than 1 year,1879
5 years or more,1164


### 10) naics

In [0]:
%sql
SELECT DISTINCT naics, COUNT(*) FROM workspace.dev_jobmkt.stg_job_postings GROUP BY 1 ORDER BY 2 DESC

naics,COUNT(*)
null,30999
Accommodation and food services,5088
Construction,3110
Retail trade,2297
Manufacturing,1909
Health care and social assistance,1443
Private households,1441
"Agriculture, forestry, fishing and hunting",1324
Transportation and warehousing,1304
"Professional, scientific and technical services",832


### 11) official_language

In [0]:
%sql
SELECT DISTINCT official_language, COUNT(*) FROM workspace.dev_jobmkt.stg_job_postings GROUP BY 1 ORDER BY 2 DESC

official_language,COUNT(*)
English,23544
*No data,19460
French,5511
English or French,4498
Bilingual,786


### 12) province_territory

In [0]:
%sql
SELECT DISTINCT province_territory, COUNT(*) FROM workspace.dev_jobmkt.stg_job_postings GROUP BY 1 ORDER BY 2 DESC

province_territory,COUNT(*)
Ontario,15947
Québec,11096
British Columbia,9441
Alberta,6549
Saskatchewan,4227
Nova Scotia,2007
Manitoba,1713
New Brunswick,1300
Newfoundland and Labrador,851
Prince Edward Island,334


### 13) city

In [0]:
%sql
SELECT DISTINCT city, COUNT(*) FROM workspace.dev_jobmkt.stg_job_postings GROUP BY 1 ORDER BY 2 DESC

city,COUNT(*)
Toronto,1892
Calgary,1870
Edmonton,1586
Vancouver,1412
Surrey,1234
Montréal,1195
Mississauga,1156
Québec,1118
Winnipeg,1100
Saskatoon,885


### 14) economic_region

In [0]:
%sql
SELECT DISTINCT economic_region, COUNT(*) FROM workspace.dev_jobmkt.stg_job_postings GROUP BY 1 ORDER BY 2 DESC

economic_region,COUNT(*)
Toronto Region,7314
Lower Mainland–Southwest Region,5869
Calgary Region,2275
Edmonton Region,2179
Montréal Region,1989
Montérégie Region,1738
Hamilton–Niagara Peninsula Region,1738
Ottawa Region,1664
Capitale-Nationale Region,1511
Vancouver Island and Coast Region,1480


### 15) work_location_postal_code

In [0]:
%sql
SELECT DISTINCT work_location_postal_code, COUNT(*) FROM workspace.dev_jobmkt.stg_job_postings GROUP BY 1 ORDER BY 2 DESC

work_location_postal_code,COUNT(*)
null,26921
G1R3X5,63
A1A5J1,37
J1K1B7,28
J2S8E3,27
V0E2S0,27
G4R4X6,24
R3C2E6,23
J4Y0G4,22
G0R3G0,21


### 16) employment_type

In [0]:
%sql
SELECT DISTINCT employment_type, COUNT(*) FROM workspace.dev_jobmkt.stg_job_postings GROUP BY 1 ORDER BY 2 DESC

employment_type,COUNT(*)
Full time,43471
Part time,7329
null,1798
Part time leading to full time,1201


### 17) employment_term

In [0]:
%sql
SELECT DISTINCT employment_term, COUNT(*) FROM workspace.dev_jobmkt.stg_job_postings GROUP BY 1 ORDER BY 2 DESC

employment_term,COUNT(*)
Permanent employment,34096
null,12098
Term or contract,4007
Seasonal employment,1806
Casual employment,1792


### 18) employment_term_weekend

In [0]:
%sql
SELECT DISTINCT employment_term_weekend, COUNT(*) FROM workspace.dev_jobmkt.stg_job_postings GROUP BY 1 ORDER BY 2 DESC

employment_term_weekend,COUNT(*)
null,30999
No,13911
Yes,8889


### 19) employment_term_telework

In [0]:
%sql
SELECT DISTINCT employment_term_telework, COUNT(*) FROM workspace.dev_jobmkt.stg_job_postings GROUP BY 1 ORDER BY 2 DESC

employment_term_telework,COUNT(*)
null,30999
No,22800


### 20) salary_condition_detail

In [0]:
%sql
SELECT DISTINCT salary_condition_detail, COUNT(*) FROM workspace.dev_jobmkt.stg_job_postings GROUP BY 1 ORDER BY 2 DESC

salary_condition_detail,COUNT(*)
$20.00 hourly,1469
$18.00 hourly,961
$21.00 hourly,876
$16.60 hourly,868
$17.60 hourly,846
$37.00 hourly,812
$25.00 hourly,791
$18.25 hourly,737
$36.00 hourly,735
$19.00 hourly,732


### 21) salary_period

In [0]:
%sql
SELECT DISTINCT salary_period, COUNT(*) FROM workspace.dev_jobmkt.stg_job_postings GROUP BY 1 ORDER BY 2 DESC

salary_period,COUNT(*)
Hour,45760
Year,7039
Month,362
Day,249
Bi-weekly,156
Week,156
null,77


### 22) salary_min

In [0]:
%sql
SELECT DISTINCT salary_min, COUNT(*) FROM workspace.dev_jobmkt.stg_job_postings GROUP BY 1 ORDER BY 2 DESC

salary_min,COUNT(*)
20.0,3170
18.0,2372
25.0,1910
17.6,1634
19.0,1612
22.0,1395
21.0,1390
17.0,1062
16.6,1027
30.0,999


### 23) salary_max

In [0]:
%sql
SELECT DISTINCT salary_max, COUNT(*) FROM workspace.dev_jobmkt.stg_job_postings GROUP BY 1 ORDER BY 2 DESC

salary_max,COUNT(*)
20.0,2445
25.0,1963
22.0,1552
18.0,1496
21.0,1401
30.0,1377
19.0,1197
17.6,1155
23.0,1027
37.0,968


### 24) hours_per

In [0]:
%sql
SELECT DISTINCT hours_per, COUNT(*) FROM workspace.dev_jobmkt.stg_job_postings GROUP BY 1 ORDER BY 2 DESC

hours_per,COUNT(*)
null,30999
Week,21732
Bi-weekly,899
Year,87
Month,82


### 25) hours_min

In [0]:
%sql
SELECT DISTINCT hours_min, COUNT(*) FROM workspace.dev_jobmkt.stg_job_postings GROUP BY 1 ORDER BY 2 DESC

hours_min,COUNT(*)
null,27529
40.0,9864
30.0,6202
35.0,4022
32.0,1445
37.5,616
20.0,406
80.0,326
15.0,231
36.0,205


### 26) hours_max

In [0]:
%sql
SELECT DISTINCT hours_max, COUNT(*) FROM workspace.dev_jobmkt.stg_job_postings GROUP BY 1 ORDER BY 2 DESC

hours_max,COUNT(*)
null,27529
40.0,14499
35.0,2989
30.0,1968
44.0,913
32.0,652
60.0,627
37.5,593
50.0,535
80.0,381


## Checking NULL ratios

In [0]:
%sql
SELECT * 
FROM(
    SELECT
        ROUND(SUM(CASE WHEN job_posting_id IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS job_posting_id,
        ROUND(SUM(CASE WHEN job_title IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS job_title,
        ROUND(SUM(CASE WHEN noc16_code IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS noc16_code,
        ROUND(SUM(CASE WHEN noc16_code_name IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS noc16_code_name,
        ROUND(SUM(CASE WHEN noc21_code IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS noc21_code,
        ROUND(SUM(CASE WHEN noc21_code_name IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS noc21_code_name,
        ROUND(SUM(CASE WHEN posting_date IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS posting_date,
        ROUND(SUM(CASE WHEN vacancy_count IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS vacancy_count,
        ROUND(SUM(CASE WHEN education_los IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS education_los,
        ROUND(SUM(CASE WHEN experience_level IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS experience_level,
        ROUND(SUM(CASE WHEN naics IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS naics,
        ROUND(SUM(CASE WHEN official_language IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS official_language,
        ROUND(SUM(CASE WHEN province_territory IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS province_territory,
        ROUND(SUM(CASE WHEN city IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS city,
        ROUND(SUM(CASE WHEN economic_region IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS economic_region,
        ROUND(SUM(CASE WHEN work_location_postal_code IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS work_location_postal_code,
        ROUND(SUM(CASE WHEN employment_type IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS employment_type,
        ROUND(SUM(CASE WHEN employment_term IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS employment_term,
        ROUND(SUM(CASE WHEN employment_term_weekend IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS employment_term_weekend,
        ROUND(SUM(CASE WHEN employment_term_telework IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS employment_term_telework,
        ROUND(SUM(CASE WHEN salary_condition_detail IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS salary_condition_detail,
        ROUND(SUM(CASE WHEN salary_period IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS salary_period,
        ROUND(SUM(CASE WHEN salary_min IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS salary_min,
        ROUND(SUM(CASE WHEN salary_max IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS salary_max,
        ROUND(SUM(CASE WHEN hours_per IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS hours_per,
        ROUND(SUM(CASE WHEN hours_min IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS hours_min,
        ROUND(SUM(CASE WHEN hours_max IS NULL THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) AS hours_max
    FROM workspace.dev_jobmkt.stg_job_postings
)
UNPIVOT (
    null_pct FOR column_name IN (
        job_posting_id, job_title, noc16_code, noc16_code_name, noc21_code, noc21_code_name, posting_date, vacancy_count, education_los, experience_level, naics, official_language, province_territory, city, economic_region, work_location_postal_code, employment_type, employment_term, employment_term_weekend, employment_term_telework, salary_condition_detail, salary_period, salary_min, salary_max, hours_per, hours_min, hours_max
    )
)
ORDER BY null_pct DESC;

column_name,null_pct
education_los,57.6
experience_level,57.6
naics,57.6
employment_term_weekend,57.6
employment_term_telework,57.6
hours_per,57.6
hours_min,51.2
hours_max,51.2
work_location_postal_code,50.0
employment_term,22.5


### Columns with High Percentage of Nulls

The columns with 50% or more than 50% null values are: education_los, experience_level, naics, employment_term_weekend,
employment_term_telework, hours_per, hours_min, hours_max, and work_location_postal_code. The next step is to check if the 57.6% observed in the columns education_los, experience_level, naics, employment_term_weekend, employment_term_telework, and hours_per is co-occurrent or independent.

In [0]:
%sql
SELECT COUNT(*) AS rows_missing_all_together
FROM workspace.dev_jobmkt.stg_job_postings
WHERE education_los IS NULL
  AND experience_level IS NULL
  AND naics IS NULL
  AND employment_term_weekend IS NULL
  AND employment_term_telework IS NULL
  AND hours_per IS NULL;

rows_missing_all_together
30999


In [0]:
%sql
SELECT COUNT(*) AS total_rows
FROM workspace.dev_jobmkt.stg_job_postings

total_rows
53799


## Column completeness: two posting "shapes"
education_los, experience_level, naics, employment_term_weekend,
employment_term_telework, and hours_per are all null together in
exactly 30,999 rows (57.6%). I confirmed this via a co-occurrence query.
This behavior can't be consider independent sparsity; it's structural, suggesting two
possible posting types (detailed vs. minimal). 

Decision: dropped education_los, experience_level, naics,
employment_term_weekend, employment_term_telework, and
work_location_postal_code from the model (too sparse to group by
usefully as dimension attributes). Kept hours_per, hours_min,
hours_max as fact measures despite the same null rate, since a
missing measure is a valid state, not a data-quality problem, 
and hours_per specifically enables more accurate salary
annualization for the ~42% of postings that have it.

Possible future addition: a posting_detail_level flag on
fct_postings (e.g. 'detailed' vs 'minimal') derived from whether
this null cluster is populated, so the two-path pattern is queryable
rather than just documented.

## List of Columns Kept for the Next Stages

* job_posting_id
* job_title
* noc16_code
* noc16_code_name 
* noc21_code 
* noc21_code_name
* posting_date
* vacancy_count
* official_language
* province_territory
* city
* economic_region
* employment_type
* employment_term
* salary_condition_detail
* salary_period
* salary_min
* salary_max
* hours_per
* hours_min
* hours_max